cmd/fleet/main.go

```diff
@@ -220,226 +219,371 @@ func NewAgentMode(cliCfg *ucfg.Config, reader io.Reader, bi build.Info, reloadab
+					a.unitRemoved(change.Unit)
+				}
+			case <-t.C:
+				// Fleet Server is the only component that gets started by Elastic Agent without an Agent ID. We loop
+				// here on interval waiting for the Elastic Agent to enroll so then the Agent ID is then set.
+				agentInfo := a.agent.AgentInfo()
+				if agentInfo != nil && agentInfo.ID != "" {
+					// Agent ID is not set for the component.
+					t.Stop()
+					err := a.reconfigure(subCtx)
+					if err != nil {
+						log.Error().Err(err)
+					}
+				}
+			}
+		}
+	}()
 
-	a.firstCfg = make(chan firstCfg)
-	a.startChan = make(chan struct{}, 1)
 	log.Info().Msg("starting communication connection back to Elastic Agent")
-	err := a.agent.Start(ctx)
+	err := a.agent.Start(subCtx)
 	if err != nil {
 		return err
 	}
 
-	// wait for the initial configuration to be sent from the
-	// Elastic Agent before starting the actual Fleet Server.
-	log.Info().Msg("waiting for Elastic Agent to send initial configuration")
-	var cfg firstCfg
-	select {
-	case <-ctx.Done():
-		return fmt.Errorf("never received initial configuration: %w", ctx.Err())
-	case cfg = <-a.firstCfg:
+	<-subCtx.Done()
+	wg.Wait()
+
+	return nil
+}
+
+// UpdateState updates the state of the message and payload.
+func (a *AgentMode) UpdateState(state client.UnitState, message string, payload map[string]interface{}) error {
+	if a.inputUnit != nil {
+		_ = a.inputUnit.UpdateState(state, message, payload)
 	}
+	if a.outputUnit != nil {
+		_ = a.outputUnit.UpdateState(state, message, payload)
+	}
+	return nil
+}
 
-	// possible that first configuration resulted in an error
-	if cfg.err != nil {
-		// unblock startChan even though there was an error
-		a.startChan <- struct{}{}
-		return cfg.err
+func (a *AgentMode) unitAdded(ctx context.Context, unit *client.Unit) error {
+	if unit.Type() == client.UnitTypeInput {
+		_, _, cfg := unit.Expected()
+		if cfg.Type != kFleetServer {
+			// not support input type
+			_ = unit.UpdateState(client.UnitStateFailed, fmt.Sprintf("%s is an unsupported input type", cfg.Type), nil)
+			return nil
+		}
+		if a.inputUnit != nil {
+			// original input unit is being stopped; swapping in this unit as the new input unit
+			_ = a.inputUnit.UpdateState(client.UnitStateStopped, kStopped, nil)
+		}
+		a.inputUnit = unit
+		if a.outputUnit == nil {
+			// waiting for output unit to really start Fleet Server
+			_ = unit.UpdateState(client.UnitStateStarting, "waiting for output unit", nil)
+			return nil
+		}
+		return a.start(ctx)
+	}
+	if unit.Type() == client.UnitTypeOutput {
+		_, _, cfg := unit.Expected()
+		if cfg.Type != kElasticsearch {
+			// not support output type
+			_ = unit.UpdateState(client.UnitStateFailed, fmt.Sprintf("%s is an unsupported output type", cfg.Type), nil)
+			return nil
+		}
+		if a.outputUnit != nil {
+			// original output unit is being stopped; swapping in this unit as the new output unit
+			_ = a.outputUnit.UpdateState(client.UnitStateStopped, kStopped, nil)
```

Comment:
What happens if this RPC call fails but the subsequent call for `Starting` succeeds? Is the `Stopped` nice to have or necessary?

Reply:
In general if an input implements the state transitions incorrectly what happens? How would they find out and know to correct it?

Reply:
Fleet Server is a very weird input, because it's the only input that gets started before a policy is every read on first launch of an Elastic Agent with a Fleet Server bootstrap. Fleet Server is the only input that will ever have to worry about this.

The `Stopped` is just more a nice to have, not really a hard requirement.